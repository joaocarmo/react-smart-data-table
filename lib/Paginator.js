// Import modules
import React from 'react'
import PropTypes from 'prop-types'
import { generatePagination, isUndefined } from './helpers/functions'
import './css/paginator.css'

const PaginatorItem = ({
  active, value, text, onPageChange,
}) => (
  <button
    className={`rsdt rsdt-paginate${active ? ' active ' : ' '}button`}
    type='button'
    onClick={(e) => onPageChange(e, { activePage: value })}
    disabled={isUndefined(value)}
  >
    {text}
  </button>
)

const Paginator = ({ activePage, totalPages, onPageChange }) => {
  const paginatorItems = generatePagination(activePage, totalPages)
  const onPageChangeFn = (
    typeof onPageChange === 'function' ? onPageChange : () => null
  )
  return (
    <div className='rsdt rsdt-paginate'>
      {paginatorItems.map(({ active, value, text }, idx) => (
        <PaginatorItem
          key={isUndefined(value) ? `ellipsis-${idx}` : text}
          active={active}
          value={value}
          text={text}
          onPageChange={onPageChangeFn}
        />
      ))}
    </div>
  )
}

/* Defines the type of data expected in each passed prop */
Paginator.propTypes = {
  activePage: PropTypes.number.isRequired,
  totalPages: PropTypes.number.isRequired,
  onPageChange: PropTypes.func,
}

PaginatorItem.propTypes = {
  active: PropTypes.bool.isRequired,
  value: PropTypes.number,
  text: PropTypes.string.isRequired,
  onPageChange: PropTypes.func.isRequired,
}

export default Paginator
