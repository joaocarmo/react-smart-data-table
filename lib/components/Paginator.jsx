import { memo, useMemo } from 'react'
import PropTypes from 'prop-types'
import PaginatorItem from './PaginatorItem'
import { generatePagination, isUndefined } from '../helpers/functions'
import '../css/paginator.css'

const areEqual = (
  { activePage: prevActivePage, totalPages: prevTotalPages },
  { activePage: nextActivePage, totalPages: nextTotalPages },
) => prevActivePage === nextActivePage && prevTotalPages === nextTotalPages

const Paginator = ({ activePage, totalPages, onPageChange }) => {
  const paginatorItems = useMemo(
    () => generatePagination(activePage, totalPages),
    [activePage, totalPages],
  )
  const onPageChangeFn = useMemo(
    () => (typeof onPageChange === 'function' ? onPageChange : () => null),
    [onPageChange],
  )

  return (
    <div className="rsdt rsdt-paginate">
      {paginatorItems.map(({ active, value, text }, idx) => (
        <PaginatorItem
          key={isUndefined(value) ? `ellipsis-${idx}` : text}
          active={active}
          value={value}
          text={text}
          onPageChange={onPageChangeFn}
        />
      ))}
    </div>
  )
}

/* Defines the type of data expected in each passed prop */
Paginator.propTypes = {
  activePage: PropTypes.number.isRequired,
  totalPages: PropTypes.number.isRequired,
  onPageChange: PropTypes.func,
}

export default memo(Paginator, areEqual)
