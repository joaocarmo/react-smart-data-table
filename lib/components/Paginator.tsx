import { memo, useMemo } from 'react'
import PropTypes from 'prop-types'
import PaginatorItem, { PageChangeFn } from './PaginatorItem'
import * as utils from '../helpers/functions'
import '../css/paginator.css'

export interface PaginatorProps {
  activePage: number
  totalPages: number
  onPageChange: PageChangeFn
}

const areEqual = (
  { activePage: prevActivePage, totalPages: prevTotalPages },
  { activePage: nextActivePage, totalPages: nextTotalPages },
) => prevActivePage === nextActivePage && prevTotalPages === nextTotalPages

const Paginator = ({
  activePage,
  totalPages,
  onPageChange,
}: PaginatorProps) => {
  const paginatorItems = useMemo(
    () => utils.generatePagination(activePage, totalPages),
    [activePage, totalPages],
  )
  const onPageChangeFn = useMemo(
    () => (typeof onPageChange === 'function' ? onPageChange : () => null),
    [onPageChange],
  )

  return (
    <div className="rsdt rsdt-paginate">
      {paginatorItems.map(({ active, value, text }, idx) => (
        <PaginatorItem
          key={utils.isUndefined(value) ? `ellipsis-${idx}` : text}
          active={active}
          value={value}
          text={text}
          onPageChange={onPageChangeFn}
        />
      ))}
    </div>
  )
}

/* Defines the type of data expected in each passed prop */
Paginator.propTypes = {
  activePage: PropTypes.number.isRequired,
  totalPages: PropTypes.number.isRequired,
  onPageChange: PropTypes.func.isRequired,
}

export default memo(Paginator, areEqual)
