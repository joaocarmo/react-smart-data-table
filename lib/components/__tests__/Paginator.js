// Test 'Paginator' component
import React from 'react'
import { mount } from 'enzyme'
import { Paginator } from '../Paginator'

test("'Paginator' component (snapshot)", () => {
  const wrap = mount(<Paginator activePage={3} totalPages={10} />)
  expect(wrap).toMatchSnapshot()
})
