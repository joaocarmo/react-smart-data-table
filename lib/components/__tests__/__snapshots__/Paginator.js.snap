// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`'Paginator' component (snapshot) 1`] = `
<Paginator
  activePage={3}
  totalPages={10}
>
  <div
    className="rsdt rsdt-paginate"
  >
    <PaginatorItem
      active={false}
      key="<<"
      onPageChange={[Function]}
      text="<<"
      value={1}
    >
      <button
        className="rsdt rsdt-paginate button"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        &lt;&lt;
      </button>
    </PaginatorItem>
    <PaginatorItem
      active={false}
      key="<"
      onPageChange={[Function]}
      text="<"
      value={2}
    >
      <button
        className="rsdt rsdt-paginate button"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        &lt;
      </button>
    </PaginatorItem>
    <PaginatorItem
      active={false}
      key="1"
      onPageChange={[Function]}
      text="1"
      value={1}
    >
      <button
        className="rsdt rsdt-paginate button"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        1
      </button>
    </PaginatorItem>
    <PaginatorItem
      active={false}
      key="2"
      onPageChange={[Function]}
      text="2"
      value={2}
    >
      <button
        className="rsdt rsdt-paginate button"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        2
      </button>
    </PaginatorItem>
    <PaginatorItem
      active={true}
      key="3"
      onPageChange={[Function]}
      text="3"
      value={3}
    >
      <button
        className="rsdt rsdt-paginate active button"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        3
      </button>
    </PaginatorItem>
    <PaginatorItem
      active={false}
      key="4"
      onPageChange={[Function]}
      text="4"
      value={4}
    >
      <button
        className="rsdt rsdt-paginate button"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        4
      </button>
    </PaginatorItem>
    <PaginatorItem
      active={false}
      key="ellipsis-6"
      onPageChange={[Function]}
      text="..."
    >
      <button
        className="rsdt rsdt-paginate button"
        disabled={true}
        onClick={[Function]}
        type="button"
      >
        ...
      </button>
    </PaginatorItem>
    <PaginatorItem
      active={false}
      key="10"
      onPageChange={[Function]}
      text="10"
      value={10}
    >
      <button
        className="rsdt rsdt-paginate button"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        10
      </button>
    </PaginatorItem>
    <PaginatorItem
      active={false}
      key=">"
      onPageChange={[Function]}
      text=">"
      value={4}
    >
      <button
        className="rsdt rsdt-paginate button"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        &gt;
      </button>
    </PaginatorItem>
    <PaginatorItem
      active={false}
      key=">>"
      onPageChange={[Function]}
      text=">>"
      value={10}
    >
      <button
        className="rsdt rsdt-paginate button"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        &gt;&gt;
      </button>
    </PaginatorItem>
  </div>
</Paginator>
`;
